File: backend/app/services/form_validation.py
from typing import Any

from app.schemas.form_template import FieldDefinition, FieldType

SELECTION_TYPES = {FieldType.SELECAO_UNICA, FieldType.SELECAO_MULTIPLA}


def parse_fields(schema: list[dict]) -> list[FieldDefinition]:
    return [FieldDefinition(**f) for f in schema]


def is_field_active(field: FieldDefinition, answers: dict[str, Any]) -> bool:
    """Um campo condicional so e considerado ativo (e portanto obrigatorio,
    se for o caso) quando a condicao e satisfeita."""
    if field.condition is None:
        return True
    return str(answers.get(field.condition.field)) == field.condition.equals


def validate_answer_values(fields: list[FieldDefinition], answers: dict[str, Any]) -> list[str]:
    """Valida tipos/opcoes das respostas presentes. Nao exige presenca de
    campos obrigatorios (ver validate_required_fields, usado so no envio)."""
    errors: list[str] = []
    fields_by_key = {f.key: f for f in fields}

    for key in answers:
        if key not in fields_by_key:
            errors.append(f"Campo desconhecido: '{key}'")

    for key, value in answers.items():
        field = fields_by_key.get(key)
        if field is None or value is None:
            continue
        if field.type == FieldType.SELECAO_UNICA:
            if value not in (field.options or []):
                errors.append(f"Campo '{key}': valor '{value}' nao esta entre as opcoes")
        elif field.type == FieldType.SELECAO_MULTIPLA:
            if not isinstance(value, list):
                errors.append(f"Campo '{key}': espera uma lista de opcoes")
            else:
                invalid = [v for v in value if v not in (field.options or [])]
                if invalid:
                    errors.append(f"Campo '{key}': valores invalidos {invalid}")
        elif field.type == FieldType.NUMERO:
            if not isinstance(value, (int, float)) or isinstance(value, bool):
                errors.append(f"Campo '{key}': deve ser numero")

    return errors


def validate_required_fields(fields: list[FieldDefinition], answers: dict[str, Any]) -> list[str]:
    errors: list[str] = []
    for field in fields:
        if not field.required or not is_field_active(field, answers):
            continue
        value = answers.get(field.key)
        if value is None or value == "" or value == []:
            errors.append(f"Campo obrigatorio nao preenchido: '{field.key}'")
    return errors


def compute_answer_diff(old: dict[str, Any], new: dict[str, Any]) -> list[dict[str, Any]]:
    changes = []
    for key in sorted(set(old) | set(new)):
        old_value = old.get(key)
        new_value = new.get(key)
        if old_value != new_value:
            changes.append({"field": key, "old_value": old_value, "new_value": new_value})
    return changes

File: mobile/lib/services/form_validation.dart
import '../models/field_definition.dart';

/// Espelha app/services/form_validation.py (backend): precisa rodar
/// localmente porque o coletor esta offline no momento de preencher e
/// enviar o RDA — nao da pra esperar o servidor validar.
List<String> requiredFieldErrors(List<FieldDefinition> fields, Map<String, dynamic> answers) {
  final errors = <String>[];
  for (final field in fields) {
    if (!field.required || !field.isActive(answers)) continue;
    final value = answers[field.key];
    final isEmpty =
        value == null ||
        (value is String && value.isEmpty) ||
        (value is List && value.isEmpty);
    if (isEmpty) {
      errors.add("Campo obrigatorio nao preenchido: '${field.label}'");
    }
  }
  return errors;
}

List<String> answerValueErrors(List<FieldDefinition> fields, Map<String, dynamic> answers) {
  final errors = <String>[];
  final byKey = {for (final f in fields) f.key: f};

  for (final key in answers.keys) {
    if (!byKey.containsKey(key)) {
      errors.add("Campo desconhecido: '$key'");
    }
  }

  for (final entry in answers.entries) {
    final field = byKey[entry.key];
    final value = entry.value;
    if (field == null || value == null) continue;

    if (field.type == FieldType.selecaoUnica) {
      if (!(field.options ?? const []).contains(value)) {
        errors.add("Campo '${field.key}': valor '$value' nao esta entre as opcoes");
      }
    } else if (field.type == FieldType.selecaoMultipla) {
      if (value is! List) {
        errors.add("Campo '${field.key}': espera uma lista de opcoes");
      } else {
        final invalid = value.where((v) => !(field.options ?? const []).contains(v)).toList();
        if (invalid.isNotEmpty) {
          errors.add("Campo '${field.key}': valores invalidos $invalid");
        }
      }
    } else if (field.type == FieldType.numero) {
      if (value is! num) {
        errors.add("Campo '${field.key}': deve ser numero");
      }
    }
  }

  return errors;
}
